### PR TITLE
AISDK-97: Add custom vocabulary client

### PR DESCRIPTION
**Pull request by Chanc3**

### CODE REVIEW

File: README.md
Comment (amikofalvy) [inline README.md]: I would refrain from using adjectives "new" or "old" to describe entities. These terms don't hold up over more than a short time window. 1 year from now, no one will know the difference between "new" or "old" naming.

Comment (amikofalvy) [reply README.md]: Strive to write timeless documentation that someone 10 years from now will be able to read without confusion.

File: src/main/java/ai/rev/speechtotext/CustomVocabulariesClient.java
Comment (amikofalvy) [inline src/main/java/ai/rev/speechtotext/CustomVocabulariesClient.java]: You could make the class description more generic. Leave it to the individual methods to describe their own purpose.

```suggestion
 * The CustomVocabularyClient object provides methods to interact with the Custom Vocabulary Api
```

File: src/main/java/ai/rev/speechtotext/ApiInterface.java
Comment (amikofalvy) [inline src/main/java/ai/rev/speechtotext/ApiInterface.java]: Since the CV API is a different api, can we define the methods in their own interface?

File: src/main/java/ai/rev/speechtotext/CustomVocabulariesClient.java
Comment (amikofalvy) [inline src/main/java/ai/rev/speechtotext/CustomVocabulariesClient.java]:
```suggestion
   * This methods makes a POST request to the /vocabularies endpoint, starts an asynchronous custom vocabulary job to
```

File: src/main/java/ai/rev/speechtotext/CustomVocabulariesClient.java
Comment (amikofalvy) [inline src/main/java/ai/rev/speechtotext/CustomVocabulariesClient.java]: Are these parameters formatted correctly? It seems strange to line break before the first parameter in order to line wrap.

File: src/main/java/ai/rev/speechtotext/CustomVocabulariesClient.java
Comment (amikofalvy) [inline src/main/java/ai/rev/speechtotext/CustomVocabulariesClient.java:54]: Should we be throwing the more specific  `IllegalArgumentException`?

File: src/main/java/ai/rev/speechtotext/ApiInterface.java
Comment (amikofalvy) [inline src/main/java/ai/rev/speechtotext/ApiInterface.java]: There is no coupling between job options and cv options. This parameter shouldn't be the same RevAiJobOptions object. There should be a separate RevAiCustomVocabularyOptions definition. Some of the properties will have similar names, but this is fine because they are separate APIs with separate options.

File: src/main/java/ai/rev/speechtotext/CustomVocabulariesClient.java
Comment (amikofalvy) [reply src/main/java/ai/rev/speechtotext/CustomVocabulariesClient.java]: The other question I have is about the ordering of the parameters. I would normally order them by importance with the optional parameters coming last: `customVocabularies`, `metadata`, `callbackUrl`.

I think that you have chosen this order so that you can override the method so that both metadata and callbackurl can be optional. I think it is only because the types of theparameters are List<CustomVocabulary>, string, string. The last two parameters would be indistinguishable if you did not have the order that you have written. Is this correct?

I worry that in the effort to create an exhaustive list of overrides we have unintentionally made the ordering of the parameters too confusing. If necessary, I would prefer to have a non-exhastive list of overrides and a more clear parameter ordering. With just a single override method, extensive customization of the parameters can be done by configuring a CustomVocabularyJobOption object with the desired properties and injecting it as a parameter to the function.

File: src/main/java/ai/rev/speechtotext/models/CustomVocabularyInformation.java
Comment (amikofalvy) [inline src/main/java/ai/rev/speechtotext/models/CustomVocabularyInformation.java]: I think we need a new object definition for CustomVocab status.

File: src/main/java/ai/rev/speechtotext/models/CustomVocabularyInformation.java
Comment (amikofalvy) [inline src/main/java/ai/rev/speechtotext/models/CustomVocabularyInformation.java]: We need cv specific failure type object too.

File: src/main/java/ai/rev/speechtotext/helpers/ClientHelper.java
Comment (amikofalvy) [inline src/main/java/ai/rev/speechtotext/helpers/ClientHelper.java:10]: Can this be used by the async api client as well?

File: src/test/java/ai/rev/speechtotext/unit/RevAiCustomVocabularyTest.java
Comment (amikofalvy) [inline src/test/java/ai/rev/speechtotext/unit/RevAiCustomVocabularyTest.java]: I think you're missing an assertion on the request body. It seems we're only checking for url and apiMethod

**Review by amikofalvy — CHANGES_REQUESTED** [review]
I left a few suggestions inline.

File: README.md
Comment (Chanc3) [reply README.md]: Funny. I pulled this from our other SDKs. I will edit and file some tickets to update the python and node readme's

File: src/main/java/ai/rev/speechtotext/CustomVocabulariesClient.java
Comment (Chanc3) [reply src/main/java/ai/rev/speechtotext/CustomVocabulariesClient.java]: Your reasoning as to the ordering is correct. I too was feeling that we should have an options object but in order to follow the other SDK implementations it requires this parameter ordering. Are we ok with deviating here?

File: src/main/java/ai/rev/speechtotext/helpers/ClientHelper.java
Comment (Chanc3) [reply src/main/java/ai/rev/speechtotext/helpers/ClientHelper.java:10]: Yes! That was the intention. Good catch.

File: README.md
Comment (jennywong2129) [inline README.md:232]: Should you mention the limits on how many phrases you can have in a CV/how many CVs can be used for a job?

File: src/main/java/ai/rev/speechtotext/models/CustomVocabularyInformation.java
Comment (jennywong2129) [inline src/main/java/ai/rev/speechtotext/models/CustomVocabularyInformation.java]: Sets*

File: README.md
Comment (Chanc3) [reply README.md:232]: I think we are not mentioning the things like limits on purpose to prevent having to update the documentation in too many places.

File: src/main/java/ai/rev/speechtotext/AsyncApiInterface.java
Comment (amikofalvy) [inline src/main/java/ai/rev/speechtotext/AsyncApiInterface.java]: Since this is public access changing the name of this interface would be a breaking change, right?

File: src/main/java/ai/rev/speechtotext/CustomVocabularyApiInterface.java
Comment (amikofalvy) [inline src/main/java/ai/rev/speechtotext/CustomVocabularyApiInterface.java:17]: I think this needs documentation like the proceeding ApiInterface: https://github.com/revdotcom/revai-java-sdk/pull/19/files#diff-bc6d97b48980ad0bde766336dd7e5d6bR25

File: src/main/java/ai/rev/speechtotext/clients/ApiClient.java
Comment (amikofalvy) [inline src/main/java/ai/rev/speechtotext/clients/ApiClient.java]: I think changing the package is another breaking change.

File: src/main/java/ai/rev/speechtotext/clients/ApiClient.java
Comment (amikofalvy) [inline src/main/java/ai/rev/speechtotext/clients/ApiClient.java]: renaming any public member is a breaking change, right?

Comment (amikofalvy) [reply src/main/java/ai/rev/speechtotext/clients/ApiClient.java]: Maybe these should be private?

File: src/main/java/ai/rev/speechtotext/AsyncApiInterface.java
Comment (Chanc3) [reply src/main/java/ai/rev/speechtotext/AsyncApiInterface.java]: yes you are right. I'll have to revert these organizational changes and renaming.

File: src/main/java/ai/rev/speechtotext/clients/ApiClient.java
Comment (Chanc3) [reply src/main/java/ai/rev/speechtotext/clients/ApiClient.java]: Unfortunately because the unit tests need the interfaces to create mock retrofit objects we can't make them private.

File: src/main/java/ai/rev/speechtotext/CustomVocabulariesClient.java
Comment (KostasRev) [inline src/main/java/ai/rev/speechtotext/CustomVocabulariesClient.java]: Given that we are doing this check for the customer, might as well check that the `accessToken` is not empty.

File: src/main/java/ai/rev/speechtotext/ApiInterface.java
Comment (KostasRev) [inline src/main/java/ai/rev/speechtotext/ApiInterface.java:58]: Curious why you indented this 4 spaces and not 2. Is this according to the style convention? We follow Google style convention right?

File: src/main/java/ai/rev/speechtotext/CustomVocabulariesClient.java
Comment (KostasRev) [inline src/main/java/ai/rev/speechtotext/CustomVocabulariesClient.java]:
```suggestion
   * This method makes a POST request to the /vocabularies endpoint and returns a {@link CustomVocabularyInformation}
   * object that provides details about the custom vocabulary submission and its progress.
```

I think it might be unnecessary to describe the "asynchronous custom vocabulary job to process the vocabularies" part which happens on the server side.

File: src/main/java/ai/rev/speechtotext/CustomVocabulariesClient.java
Comment (KostasRev) [inline src/main/java/ai/rev/speechtotext/CustomVocabulariesClient.java]: Similarly here, we can check if the list is empty. Update the docstring as well if you make that change.

File: src/main/java/ai/rev/speechtotext/CustomVocabulariesClient.java
Comment (KostasRev) [reply src/main/java/ai/rev/speechtotext/CustomVocabulariesClient.java]: I think it is preferable to use the options object rather than mirror the other SDKs; it is better code practice. I think the reason we did not go down that path in the other SDKs is that we would have introduced a breaking change.

File: src/main/java/ai/rev/speechtotext/models/vocabulary/CustomVocabularyOptions.java
Comment (KostasRev) [inline src/main/java/ai/rev/speechtotext/models/vocabulary/CustomVocabularyOptions.java]: I think it would be better to name this class `SubmitCustomVocabularyOptions`

File: src/test/java/ai/rev/speechtotext/unit/RevAiCustomVocabularyTest.java
Comment (KostasRev) [inline src/test/java/ai/rev/speechtotext/unit/RevAiCustomVocabularyTest.java]: Hmm, is it conventional to put the method name on separate line from its access modifier and return type? It looks weird. Test names are expected to be long so I wouldn't worry about line length minimization if convention is not at play here.

Comment (KostasRev) [reply src/test/java/ai/rev/speechtotext/unit/RevAiCustomVocabularyTest.java]:
```suggestion
  public void
      submitCustomVocabulary_WhenVocabularyCallbackAndMetadataAreSpecified_ReturnsACustomVocabularyInformation()
```

File: src/test/java/ai/rev/speechtotext/unit/RevAiCustomVocabularyTest.java
Comment (KostasRev) [inline src/test/java/ai/rev/speechtotext/unit/RevAiCustomVocabularyTest.java]: It would be better to extract the large `setup` blocks of code into private helper methods if possible. With that approach you can also have all the private helper method calls be packed at the top of the method together which helps keep the test easy to read with a clear visual separation of the "arrange/act/assert" sections.

File: src/test/java/ai/rev/speechtotext/unit/RevAiCustomVocabularyTest.java
Comment (KostasRev) [reply src/test/java/ai/rev/speechtotext/unit/RevAiCustomVocabularyTest.java]: Another note is that you can drop the "When" in the condition. Rev convention in tests is "method_condition_result" so I think the "when" is always understood by implication.

File: src/main/java/ai/rev/speechtotext/clients/ApiClient.java
Comment (KostasRev) [reply src/main/java/ai/rev/speechtotext/clients/ApiClient.java]: Can they be protected?

File: src/main/java/ai/rev/speechtotext/CustomVocabulariesClient.java
Comment (KostasRev) [inline src/main/java/ai/rev/speechtotext/CustomVocabulariesClient.java]: Can check if string is empty here as well.

**Review by KostasRev — CHANGES_REQUESTED** [review]
Lookin good, I made some suggestions. Please manually test the client as well after all the changes.

File: src/main/java/ai/rev/speechtotext/clients/ApiClient.java
Comment (Chanc3) [reply src/main/java/ai/rev/speechtotext/clients/ApiClient.java]: Unfortunately you cannot declare the class protected. But even if it were, that would make this class accessible only within the same package which would prevent the unit tests from using it :(

File: src/main/java/ai/rev/speechtotext/ApiInterface.java
Comment (Chanc3) [reply src/main/java/ai/rev/speechtotext/ApiInterface.java:58]: Yup this is the [google style convention](https://google.github.io/styleguide/javaguide.html#s4.5.2-line-wrapping-indent).

File: src/test/java/ai/rev/speechtotext/unit/RevAiCustomVocabularyTest.java
Comment (Chanc3) [reply src/test/java/ai/rev/speechtotext/unit/RevAiCustomVocabularyTest.java]: This line wrapping here follows the google format. The only exceptions made are listed out [here](https://google.github.io/styleguide/javaguide.html#s4.4-column-limit).

I can drop the when and will do so for all our other tests as well.

File: src/main/java/ai/rev/speechtotext/models/vocabulary/CustomVocabularyOptions.java
Comment (Chanc3) [reply src/main/java/ai/rev/speechtotext/models/vocabulary/CustomVocabularyOptions.java]: I'm torn here. That name sounds like you are submitting options. Since the `CustomVocabulary` is required the `Options` part sounds misleading.

What about something like `CustomVocabularyContents`? What we are submitting is the contents of the vocabulary...

Comment (KostasRev) [reply src/main/java/ai/rev/speechtotext/models/vocabulary/CustomVocabularyOptions.java]: Haha I see your point. I was trying alternative names because I think `CustomVocabularyOptions` sounds like it is just the `Options` part rather than also the `CustomVocabulary` part (ie the phrases) which are core to the submission. 

What if we call it `CustomVocabularySubmission`? That makes it clear that these are not just options.

File: src/main/java/ai/rev/speechtotext/ApiClient.java
Comment (KostasRev) [inline src/main/java/ai/rev/speechtotext/ApiClient.java]: All of these emptiness checks should be `||` rather than `&&`. The purpose is to verify that the accessToken given is nonempty or null.

Please change this and all the other `&& object.isEmpty()` checks.

File: src/test/java/ai/rev/speechtotext/unit/RevAiCustomVocabularyTest.java
Comment (KostasRev) [inline src/test/java/ai/rev/speechtotext/unit/RevAiCustomVocabularyTest.java:105]: Nice.

File: src/test/java/ai/rev/speechtotext/integration/CustomVocabulariesTest.java
Comment (KostasRev) [inline src/test/java/ai/rev/speechtotext/integration/CustomVocabulariesTest.java]: Why is callback not included in this test? I would add it here.

Also I would separate the `customVocabularyInformation.getStatus()` out of the assert statement, please do the same other tests that do this as well. It is useful to have 3 visual blocks in tests:

```
     // Arrange
     CustomVocabularyInformation customVocabularyInformation = submitCustomVocabulary(); 

     // Act
     status = customVocabularyInformation.getStatus()

    // Assert ...
```

**Review by KostasRev — CHANGES_REQUESTED** [review]
One last set of changes and I think we can ship afterwards! 

Most notably the `&&` to `||` for any place that is checking for emptiness.

File: src/main/java/ai/rev/speechtotext/ApiClient.java
Comment (Chanc3) [reply src/main/java/ai/rev/speechtotext/ApiClient.java]: oh man good catch. I don't know what I was thinking.

File: src/test/java/ai/rev/speechtotext/integration/CustomVocabulariesTest.java
Comment (Chanc3) [reply src/test/java/ai/rev/speechtotext/integration/CustomVocabulariesTest.java]: The callback wasn't included because the integration test really sends a request and I don't have a valid callback url to pass in. This same logic applied to the `SubmitJobTest` integration tests. Should we just add something like `https://test.test.test`?

Are you sure about the blocks? I feel like this is a style preference and adds some unnecessary assignment. I feel like because the everything is named well it's easy to understand and reads well. `Assert that custom vocabulary information status is equal to the custom vocabulary status in_progress`, no?

Also blocks like:
```
for (CustomVocabularyInformation customVocabulary : customVocabularies) {
      assertThat(customVocabulary.getId()).as("Vocabulary Id").isNotNull();
      assertThat(customVocabulary.getCreatedOn()).as("Created on").isNotNull();
      assertThat(customVocabulary.getStatus()).as("Status").isNotNull();
      if (customVocabulary.getStatus() == CustomVocabularyStatus.FAILED) {
        assertThat(customVocabulary.getFailure()).as("Failure").isNotNull();
        assertThat(customVocabulary.getFailureDetail()).as("Failure detail").isNotNull();
      } else if (customVocabulary.getStatus() == CustomVocabularyStatus.COMPLETE) {
        assertThat(customVocabulary.getCompletedOn()).as("Completed on").isNotNull();
      }
    }
```

Will become
```
for (CustomVocabularyInformation customVocabulary : customVocabularies) {
    String id = customVocabulary.getId();
    String createdOn = customVocabulary.getCreatedOn();
    String completedOn = customVocabulary.getCompletedOn();
    CustomVocabularyStatus status = customVocabulary.getStatus();
    CustomVocabularyFailureType failure = customVocabulary.getFailure();
    String failureDetail = customVocabulary.getFailureDetail();

    assertThat(id).as("Vocabulary Id").isNotNull();
    assertThat(createdOn).as("Created on").isNotNull();
    assertThat(status).as("Status").isNotNull();
    if (status == CustomVocabularyStatus.FAILED) {
      assertThat(failure).as("Failure").isNotNull();
      assertThat(failureDetail).as("Failure detail").isNotNull();
    } else if (status == CustomVocabularyStatus.COMPLETE) {
      assertThat(completedOn).as("Completed on").isNotNull();
    }
  }
```

In these two cases I'm not really seeing a big difference but if you think this improves readability I'll make the changes everywhere. 

Will we also need to do it in other SDKs? What about tests in node like https://github.com/revdotcom/revai-node-sdk/blob/develop/test/integration/test/custom_vocabularies.test.js

Comment (Chanc3) [reply src/test/java/ai/rev/speechtotext/integration/CustomVocabulariesTest.java]: This request was clarified over a zoom and changes have been implemented.

File: src/main/java/ai/rev/speechtotext/CustomVocabulariesClient.java
Comment (KostasRev) [inline src/main/java/ai/rev/speechtotext/CustomVocabulariesClient.java]: Couple of minor things - "A object" -> "An object"

Also, I would not mention the optional fields in the docstring, this list would keep growing for one thing everytime we add a new optional parameter. I think it would suffice to say something along the lines of "An object that contains the custom vocabularies as well as some optional parameters"

Also would it help to add a {@link CustomVocabularySubmission}?

File: src/test/java/ai/rev/speechtotext/integration/CustomVocabulariesTest.java
Comment (KostasRev) [inline src/test/java/ai/rev/speechtotext/integration/CustomVocabulariesTest.java]:
```suggestion
      submitCustomVocabulary_VocabAndAllOptionsAreIncluded_ReturnCustomVocabularyInformationInProgress() {
```

Need to rename the test to indicate that all options are being specified, feel free to use an alternate phrasing to my suggestion.

File: src/test/java/ai/rev/speechtotext/integration/CustomVocabulariesTest.java
Comment (KostasRev) [inline src/test/java/ai/rev/speechtotext/integration/CustomVocabulariesTest.java:88]: Minor - is this not null check needed given that the status is checked in the next line?

File: src/test/java/ai/rev/speechtotext/integration/CustomVocabulariesTest.java
Comment (KostasRev) [inline src/test/java/ai/rev/speechtotext/integration/CustomVocabulariesTest.java]: I wonder if there is utility to check that the failure is a member of `CustomVocabularyFailureType` 🤔

File: src/test/java/ai/rev/speechtotext/integration/CustomVocabulariesTest.java
Comment (KostasRev) [inline src/test/java/ai/rev/speechtotext/integration/CustomVocabulariesTest.java:85]: Good DRYing 😎

File: src/test/java/ai/rev/speechtotext/unit/AssertHelper.java
Comment (KostasRev) [inline src/test/java/ai/rev/speechtotext/unit/AssertHelper.java:36]: Very nice!

**Review by KostasRev — APPROVED** [review]
Looks good to me, and the tests _especially_ so!

But I left some minor comments, please address those, and make sure you've tested all clients that were altered and that they behave as expected and :shipit:

File: src/test/java/ai/rev/speechtotext/integration/CustomVocabulariesTest.java
Comment (Chanc3) [reply src/test/java/ai/rev/speechtotext/integration/CustomVocabulariesTest.java:88]: I felt it was still needed because is should not be null and the next two if statements only check if it is failed or completed. This will ensure that it is always set to something.

**Pull request merged**
